### ISSUE TITLE
aws_volume_attachment Error waiting for Volume

### ISSUE THREAD

**Issue by jrobison-sb** [hashicorp/terraform-provider-aws#4770]
### Terraform Version

Terraform v0.11.7
+ provider.aws v1.20.0

### Affected Resource(s)

aws_volume_attachment

### Actual Behavior

When running a `terraform destroy`, in which you want to destroy an `aws_instance` and it's `aws_volume_attachment`, terraform (or the AWS API) will attempt to destroy the volume attachment first, then then instance second. This often fails because `Error waiting for Volume (vol-010b3d979b1027867) to detach from Instance`.

### Expected Behavior

If an `aws_volume_attachment` was going to be destroyed, and the associated `aws_instance` was also going to be destroyed, it would make sense to destroy the instance first and then the `aws_volume_attachment` second, which would never run any risk of timing out.

### Steps to Reproduce

1. Create an `aws_ebs_volume`, an `aws_instance`, and a `aws_volume_attachment`.
2. `terraform destroy`

**Comment by AranelSurion** [comment]
I can confirm this for our infra too. On latest Terraform, destroying aws_volume_attachment and aws_instance together fails with this timeout error.

**Comment by alexaandru** [comment]
Running into this as well.

**Comment by Hashfyre** [comment]
```
➜ tfa
data.aws_availability_zones.az: Refreshing state...
data.aws_route53_zone.int-orgaws: Refreshing state...
data.aws_route53_zone.orgaws: Refreshing state...
data.aws_vpc.stage-zoomcar: Refreshing state...
data.aws_ami.stage-mongodb: Refreshing state...
data.aws_subnet.stage-mongodb: Refreshing state...
data.aws_security_group.stage-mongodb: Refreshing state...
aws_instance.stage-mongodb: Refreshing state... (ID: i-xxxxxxxxxxxxxxxxxxx)
aws_route53_record.stage-mongodb-int: Refreshing state... (ID: ACCID-stage-mongodb-int_A)
aws_ebs_volume.stage-mongodb-volume-a: Refreshing state... (ID: vol-04c99a968b142f730)
aws_volume_attachment.stage-mongodb-volume-attachment-a: Refreshing state... (ID: vai-385500658)

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  - aws_ebs_volume.stage-mongodb-volume-a

  - aws_instance.stage-mongodb

  - aws_route53_record.stage-mongodb-int

  - aws_volume_attachment.stage-mongodb-volume-attachment-a


Plan: 0 to add, 0 to change, 4 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

aws_volume_attachment.stage-mongodb-volume-attachment-a: Destroying... (ID: vai-xxxxxxxxxxx)
aws_route53_record.stage-mongodb-int: Destroying... (ID: ACCID-stage-mongodb-int_A)
aws_volume_attachment.stage-mongodb-volume-attachment-a: Still destroying... (ID: vai-385500658, 10s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 10s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 20s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 30s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 40s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 50s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 1m0s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 1m10s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 1m20s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 1m30s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 1m40s elapsed)
aws_route53_record.stage-mongodb-int: Destruction complete after 1m40s

Error: Error applying plan:

1 error(s) occurred:

* aws_volume_attachment.stage-mongodb-volume-attachment-a (destroy): 1 error(s) occurred:

* aws_volume_attachment.stage-mongodb-volume-attachment-a: Error waiting for Volume (vol-xxxxxxxxxxxxx) to detach from Instance: i-xxxxxxxxxxxxx

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

**Comment by Hashfyre** [comment]
```
➜ tfa
data.aws_availability_zones.az: Refreshing state...
data.aws_route53_zone.int-orgaws: Refreshing state...
data.aws_route53_zone.orgaws: Refreshing state...
data.aws_vpc.stage-org: Refreshing state...
data.aws_ami.stage-mongodb: Refreshing state...
data.aws_subnet.stage-mongodb: Refreshing state...
data.aws_security_group.stage-mongodb: Refreshing state...
aws_instance.stage-mongodb: Refreshing state... (ID: i-xxxxxxxxxxxxxxxxxxx)
aws_route53_record.stage-mongodb-int: Refreshing state... (ID: ACCID-stage-mongodb-int_A)
aws_ebs_volume.stage-mongodb-volume-a: Refreshing state... (ID: vol-xxxxxxxxxxx)
aws_volume_attachment.stage-mongodb-volume-attachment-a: Refreshing state... (ID: vai-xxxxxxxxx)

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  - aws_ebs_volume.stage-mongodb-volume-a

  - aws_instance.stage-mongodb

  - aws_route53_record.stage-mongodb-int

  - aws_volume_attachment.stage-mongodb-volume-attachment-a


Plan: 0 to add, 0 to change, 4 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

aws_volume_attachment.stage-mongodb-volume-attachment-a: Destroying... (ID: vai-xxxxxxxxxxx)
aws_route53_record.stage-mongodb-int: Destroying... (ID: ACCID-stage-mongodb-int_A)
aws_volume_attachment.stage-mongodb-volume-attachment-a: Still destroying... (ID: vai-385500658, 10s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 10s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 20s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 30s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 40s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 50s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 1m0s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 1m10s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 1m20s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 1m30s elapsed)
aws_route53_record.stage-mongodb-int: Still destroying... (ID: ACCID-stage-mongodb-int_A, 1m40s elapsed)
aws_route53_record.stage-mongodb-int: Destruction complete after 1m40s

Error: Error applying plan:

1 error(s) occurred:

* aws_volume_attachment.stage-mongodb-volume-attachment-a (destroy): 1 error(s) occurred:

* aws_volume_attachment.stage-mongodb-volume-attachment-a: Error waiting for Volume (vol-xxxxxxxxxxxxx) to detach from Instance: i-xxxxxxxxxxxxx

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

**Comment by mikaelforeseeti** [comment]
I ran into this today, is there a workaround? I tried running `terraform destroy --target=module.machines.ep_instance` but it still tries to destroy the volume_attachment object first.

**Comment by paulrowlands** [comment]
FYI, I ran into this when implementing Windows servers with MSSQL.

I found that adding a null_resource which has a depends_on for the instance and a remote-exec provisioner which includes a 'when = "destroy"' option and executes a Powershell command stop the MSSQLSERVER service on the host.

After this, terraform destroy will destroy the attachment.

You could replace the MSSQL service with relevant shell commands to stop services, unmount disks, etc.

**Comment by yum-dev** [comment]
Even I'm facing this issue. 

Terraform v0.12.1
+ provider.aws v2.14.0

**Comment by DrHashi** [comment]
Terraform v0.12.3
+ provider.aws v2.18.0 

Still happening

**Comment by R-Ali** [comment]
I have the same issue

Terraform v0.12.5
+ provider.aws v2.20.0

**Comment by twarnock2w** [comment]
We are facing this issue as well.  

Terraform v0.12.5
provider.aws v2.22

**Comment by AndreiF8** [comment]
Facing same issue

Terraform v0.11.14
+ provider.aws v2.25.0

**Comment by pitthecat** [comment]
We have the same problem 

Terraform 0.12.7 and provider AWS 2.25.0

**Comment by yunlab** [comment]
same here
Terraform v0.12.9
+ provider.aws v2.29.0

**Comment by RiceAndMeet** [comment]
Facing the same issue here, we worked around it by including <code>skip_destroy = true</code> in resource <code>aws_volume_attachment</code>. 

According to the documentation, setting <code>skip_destroy = true</code> will remove resource  <code>aws_volume_attachment</code> from the state file, and this will unblock <code>terraform destroy </code> operation.

**Comment by hehnope** [comment]
Can confirm this bug as well:

```
Error: Error waiting for Volume (vol-xxx) to detach from Instance: i-xxx
```

**Comment by kreempuff** [comment]
Hey all, Happy Holidays. 
What fixed it for me was the first line [here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-detaching-volume.html):

> You can detach an Amazon EBS volume from an instance explicitly or by terminating the instance. However, if the instance is running, you **must first unmount the volume** from the instance. 

If the volume is still mounted on the instance, it will never detach (it even shows gray in the AWS console, so terraform may be able to help us here). Debugging terraform with `export TF_LOG=TRACE` showed that the api returned a `200 OK` for DetachVolume api call, but subsequent calls to DescribeVolume showed the volume as busy. 

After I unmounted the volume, terraform destroyed the volume attachment successfully.

**Comment by twarnock** [comment]
If we're listing workarounds, shutting down the instance before the destroy will work as well, since the detach will work since the volume would not be "busy" at that point.

**Comment by edsale** [comment]
Same problem here.  Only been a bug 19 months.  Could we get Terraform to STOP instances with attached storage, DETACH/DELETE all volumes and then TERMINATE the instance?  Seems like a fundamental function fails often in this circumstance.

**Comment by yogeshsajanikar** [comment]
Probably there should be a *provision* to run a script just before volume is detached.

**Comment by mixe3y** [comment]
+1 here
Terraform version: 0.12.25
AWS Provider version: 2.65.0

**Comment by gittur** [comment]
another day, another issue with super basic stuff expected from TF. No response from TF also. 

For the record: 
- Terraform v0.12.26

Issue persists.

**Comment by onezeroecho** [comment]
Got this issue too.
```
Terraform v0.12.28
+ provider.aws v2.63.0
```

**Comment by shrilesh1986** [comment]
Encountered Same issue, have to manually unmount to volume and then proceed with terraform destroy.

**Comment by stijnopheide** [comment]
persists in

**Comment by Jouo** [comment]
Problem still here in Terraform **v0.13.4**

**Comment by juloi** [comment]
Same problem!!! 

```
terraform v0.13.2
provider.aws  ~> 3.0
```

**Comment by assinnata** [comment]
Same issue for me

![image](https://user-images.githubusercontent.com/7115007/103345895-e013b280-4a92-11eb-8f60-d306eb174269.png)

**Comment by maciejjakub** [comment]
Same issue. We're using Terraform v0.13.4.

**Comment by dstamen** [comment]
still hitting this with terraform 1.0.2 and AWS provider 3.5. stopping the instance first allowed me to destroy.

Terraform v1.0.2
on darwin_amd64
+ provider registry.terraform.io/hashicorp/aws v3.50.0

**Comment by Illutax** [comment]
Couldn't detatch volume, not even with EC2 Management Console. The volume attachment was in state detaching. I manually stopped the instance, then the attachment was successfully detached.
After that i started the instance again and the terraform apply ran successfully.

Was using Terraform v0.12.31

**Comment by RawthiL** [comment]
Same here, using:
- Terraform v1.0.5
- hashicorp/aws  3.24.1
- aws-cli/2.2.16 Python/3.8.8 Linux/5.11.0-27-generic exe/x86_64.ubuntu.20 prompt/off  (I'm not sure if this is relevant)

If the volume is mounted the destruction fails, however if I unmount the volume then the attachment destruction, and later the instance destruction, both succeed. While this is logical, I would expect terraform to first destroy the instance (unmounting the volume in the process) and then destroy the volume attachment.

**Comment by Bdoom** [comment]
I'm also having this issue in terraform 1.0.7

**Comment by YakDriver** [comment]
Thank you for reporting this and your continued interest. This issue should be resolved with the feature added in #21144. If that does not resolve the issue, please open a new issue and reference this one.
